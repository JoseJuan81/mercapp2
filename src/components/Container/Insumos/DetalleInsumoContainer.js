import React from 'react';
import { useSelector } from 'react-redux';

import { PaginaDetalleInsumo } from '../../../Pages/Insumos/PaginaDetalleInsumo';
import { PaginaLoadingDetalleInsumo } from '../../../Pages/loading/PaginaLoadingDetalleInsumo';

import { DetalleInsumoMenuMobile } from '../../Menus/DetalleInsumoMenuMobile';

export const DetalleInsumoContainer = () => {

    // ===== STORE =====
    const { loading } = useSelector( store => store.loading );
    
    return (
        <div>
            {loading
            ? <PaginaLoadingDetalleInsumo />
            : <PaginaDetalleInsumo />}

            <DetalleInsumoMenuMobile />
        </div>
    )
}
