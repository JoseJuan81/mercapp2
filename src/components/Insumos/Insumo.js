import React from 'react';

const InsumoEtiquetas = ({ labels, checked }) => {
    return (
        <ul className="flex flex-wrap mt-2">

            {
                labels.map(la => (
                    <li
                        className={`
                            bg-white
                            text-xs ${ checked ? 'text-lime-700' : 'text-warmGray-800' }
                            mr-1 mb-1 p-1
                        `}
                    >{ la }</li>
                ))
            }

        </ul>
    )
}

export const Insumo = ({ title, price, currency, labels, checked }) => {

    return (
        <div
            className={`
                duration-200
                rounded-lg ${checked && 'shadow-xl'}
                border border-solid ${checked ? 'border-lime-400' : 'border-warmGray-300'}
            `}
        >

            <dl className="flex flex-auto p-2 mb-2">

                <div className="flex flex-auto">
                    <dt
                        className={`
                            duration-200
                            flex-auto
                            text-2xl font-bold ${checked ? 'text-lime-600' : 'text-warmGray-400'}
                        `}
                    >
                        { title }
                    </dt>

                    <dt
                        className={`
                            text-2xl ${checked ? 'text-lime-400' : 'text-warmGray-400'}
                            ${checked && 'font-bold'}
                        `}
                    >
                        <small
                            className="text-xs font-light"
                        >
                            { currency }
                        </small>
                        <span>{ price }</span>
                    </dt>
                </div>

                <div className="ml-4">...</div>

            </dl>

            <div
                className={`
                    rounded-br-lg rounded-bl-lg
                    ${ checked ? 'bg-lime-100' : 'bg-warmGray-100' }
                    px-2 py-2
                `}
            >
                <InsumoEtiquetas
                    labels={ labels }
                    checked={ checked }
                />
            </div>
            
        </div>
    )
}
