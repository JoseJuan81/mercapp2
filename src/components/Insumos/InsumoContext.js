import { createContext } from 'react';

export const InsumoContext = createContext(null);