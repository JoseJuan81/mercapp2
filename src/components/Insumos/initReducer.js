import { decide, equality, find, reduce } from 'functionallibrary';

import { data as dataInsumos } from '../../fakeData/insumos.json';

import { getFromLocalStorage } from '../../helper/localStorage';
import { alphabeticSorting } from '../../helper/alphabeticSorting';

export const initReducer = () => {
    const dataSelected = getFromLocalStorage('selected-insumos');

    if (!dataSelected || dataSelected.length === 0) {
        return alphabeticSorting(dataInsumos);
    }
    
    const reduceManage = (list, item) => {

        const isChecked = find( equality('title', item.title), dataSelected );

        const addChecked = () => list.concat(isChecked);
        const addNotChecked = (el) => list.concat(el);

        return decide(!!isChecked, addChecked, addNotChecked, item);

    }

    const updatedInsumos = reduce(reduceManage, [], dataInsumos);
    return alphabeticSorting(updatedInsumos);
}
