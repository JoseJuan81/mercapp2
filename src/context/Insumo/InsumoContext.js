import { createContext } from 'react';

/**
 * Crear contexto de Insumo
 */
export const InsumoContext = createContext(null);
