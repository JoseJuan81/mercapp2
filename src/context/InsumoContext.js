import { createContext, useEffect, useMemo, useState } from 'react';
import { removeItemFromArrayByProp } from 'functionallibrary';

import { useIndexDB } from '../hooks/useIndexDB';

import { getFromLocalStorage, setInLocalStorage } from '../helper/localStorage';
import {
    initialInsumos,
    onFilter,
    onSearch,
    updateItemInArrayById,
    selectAllInsumos,
    unSelectAllInsumos,
    updateTotal,
    updatingQuantity,
} from '../helper/helperInsumoContext';

/**
 * Crear contexto de Insumo
 */
export const InsumoContext = createContext(null);

const dataSelected = getFromLocalStorage('selected-insumos') || [];

/**
 * @description funcion wrapper del contexto
 * @param {ReactComponent} children
 */
export const InsumoStore = ({ children }) => {

    // obtener insumos de indexDB
    const {
        deleteInsumo,
        insumos: insumosInLocalBD,
        setNewInsumo
    } = useIndexDB();


    // Ordenar insumos provenientes de indexDB y actualizarlos con los seleccionados
    const initInsumos = useMemo( () => initialInsumos( insumosInLocalBD ), [insumosInLocalBD] );

    const [insumos, setInsumos] = useState(initInsumos);
    const [selectedInsumos, setSelectedInsumos] = useState(dataSelected);
    const [total, setTotal] = useState(0);


    /**
     * @description: seleccionar o deseleccionar un insumo
     * @param {object} newInsumo - insumo
     */
    const toogleCheck = (newInsumo) => {

        // Actualizar insumos
        setInsumos(
            updateItemInArrayById(insumos, newInsumo)
        )

        // actualizar insumos seleccionados
        const updatedSelectedInsumos = newInsumo.checked
            ? [newInsumo, ...selectedInsumos]
            : removeItemFromArrayByProp('id', newInsumo.id, selectedInsumos);
        
        setSelectedInsumos( [...updatedSelectedInsumos] );
        setInLocalStorage('selected-insumos', updatedSelectedInsumos);


    }

    /**
     * @description: actualiza cantidad del insumo
     * @param {string} id 
     * @param {number} quantity 
     */
    const updateQuantityInSelectedInsumo = (id, quantity) => {

        setSelectedInsumos(
            updatingQuantity( id, quantity, selectedInsumos )
        );

    }

    /**
     * @description: selecciona todos los insumos
     * @param {array} insumos - arreglo de insumos
     */
    const selectingAllInsumos = (insumos) => {

        const allSelected = selectAllInsumos(insumos);

        setInsumos( allSelected );
        setSelectedInsumos([...allSelected]);
        setInLocalStorage('selected-insumos', allSelected);
    }
    
    /**
     * @description: deselecciona todos los insumos
     * @param {array} insumos - arreglo de insumos
     */
    const unSelectingAllInsumos = () => {
        
        const allUnSelected = unSelectAllInsumos( initialInsumos() );

        setInsumos( allUnSelected );
        setSelectedInsumos( [] );
        setInLocalStorage( 'selected-insumos', [] );
    };

    /**
     * @description: Busca insumos usando el titulo
     * @param {string} searchVal - criterio de busqueda
     */
    const searchingInsumos = (searchVal) => {

        if(searchVal === '' || searchVal === null || searchVal === undefined) {

            setInsumos( initialInsumos() );
        } else {

            const result = onSearch( insumos, searchVal );
            setInsumos( result );
        }

    }

    /**
     * @description: Busca insumos usando los labels
     * @param {string} searchVal - criterio de buscqueda
     */
    const filteringInsumos = (searchVal) => {

        if(searchVal === '' || searchVal === null || searchVal === undefined) {

            setInsumos( initialInsumos() );
        } else {

            const result = onFilter( insumos, searchVal );
            setInsumos( result );
        }

    }

    const addingNewInsumo = async (newInsumo) => {

        try {

            await setNewInsumo( newInsumo );
            setInsumos( i => [{ ...newInsumo }, ...i] );

        } catch(err) {
            console.log('Erro al crear insumo', err);
        }

    }

    const deletingInsumo = async (insumoId) => {

        await deleteInsumo(insumoId);

        setInsumos(
            removeItemFromArrayByProp('id', insumoId, insumos)
        );
    }

    // Total de insumos seleccionados => price * quantity
    useEffect( () => {

        setTotal(
            updateTotal( selectedInsumos )
        );
        
    }, [selectedInsumos])

    useEffect( () => {

        setInsumos(initInsumos);

    }, [initInsumos])

    const insumoContextProps = {
        addingNewInsumo,
        deletingInsumo,
        selectingAllInsumos,
        filteringInsumos,
        insumos,
        unSelectingAllInsumos,
        searchingInsumos,
        selectedInsumos,
        total,
        toogleCheck,
        updateQuantityInSelectedInsumo,
    }

    return (
        <InsumoContext.Provider
            value={ insumoContextProps }
        >
            { children }

        </InsumoContext.Provider>
    )
}