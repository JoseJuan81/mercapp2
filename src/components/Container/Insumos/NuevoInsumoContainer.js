import React from 'react';

import { PaginaNuevoInsumo } from '../../../Pages/Insumos/PaginaNuevoInsumo';
import { PaginaLoadingPaginaNuevoInsumo } from '../../../Pages/loading/PaginaLoadingPaginaNuevoInsumo';

import { NuevoInsumoMenuMobile } from '../../Menus/NuevoInsumoMenuMobile';

export const NuevoInsumoContainer = () => {

    return (
        <div className="layout__page">
            <PaginaNuevoInsumo />

            <NuevoInsumoMenuMobile />
        </div>
    )
}
