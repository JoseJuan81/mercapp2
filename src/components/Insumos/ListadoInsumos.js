import React from 'react';
import { Insumo } from './Insumo';

export const ListadoInsumos = ({ insumosState, handleClickOnInsumo }) => {
    return (
        <ul className="">
            {
                insumosState.map((i, index) => (

                    <li
                        className="mb-3"
                        onClick={ () => handleClickOnInsumo(i) }
                    >
                        <Insumo
                            key={ `${i.title}-${index}`}
                            { ...i }
                        />
                    </li>

                ))
            }
        </ul>
    )
}
