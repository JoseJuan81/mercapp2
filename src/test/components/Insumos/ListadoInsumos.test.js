import { mount } from 'enzyme';
import { ListadoInsumos } from '../../../components/Insumos/ListadoInsumos';
import { InsumoContext } from '../../../context/Insumo/InsumoContext';

const dataFake = [
    {
        id: '82b4791a4e7cb', title: "Aceite de Oliva", price: 20, currency: "S/.", labels: ["500ml", "Carbonell", "metro", "2 und"]
    },
    {
        id: '972de0224af17', title: "Harina Pan", price: 5.50, currency: "S/.", labels: ["1 Kg", "PAN", "San Felipe", "2 und"]
    },
]

describe('Pruebas sobre archivo ListadoInsumos.js', () => {

    const wrapper = mount(
        <InsumoContext.Provider
            value={{
                insumos: []
            }}
        >
            <ListadoInsumos/>
        </InsumoContext.Provider>
    )

    test('Se debe renderizar correctamente <ListadoInsumos />', () => {

        expect( wrapper ).toMatchSnapshot();

    })

    test('Se debe rernderizar "Debe agregrar insumos :)" cuando insumoState = []', () => {
        const h1 = wrapper.find('h1');
        expect( h1.text().trim() ).toBe( 'Debe agregrar insumos :)' );

    })

    test('Se debe renderizar Insumo cuando insumoState != []', () => {

        const wrapper = mount(
            <InsumoContext.Provider
                value={{
                    insumos: dataFake
                }}
            >
            <ListadoInsumos />
        </InsumoContext.Provider>
        )
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('li.mb-3').length ).toBe( dataFake.length );
    })

    test('Se debe llamar mockHandleClickOnInsumo al hacer click', () => {

        const wrapper = mount(
            <InsumoContext.Provider
                value={{
                    insumos: dataFake
                }}
            >
                <ListadoInsumos/>
            </InsumoContext.Provider>
        )

        const li = wrapper.find('li').at(0);
        li.simulate('click', dataFake[0]);

        // expect( mockHandleClickOnInsumo ).toHaveBeenCalled();
    })
})