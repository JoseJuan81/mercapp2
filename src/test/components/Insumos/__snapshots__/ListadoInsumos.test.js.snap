// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre archivo ListadoInsumos.js Se debe renderizar Insumo cuando insumoState != [] 1`] = `
<ul
  className=""
>
  <li
    className="mb-3"
    key="Aceite de Oliva-0-function random() { [native code] }"
  >
    <div
      className="
                duration-200
                rounded-lg undefined
                border border-solid border-warmGray-300
            "
      onClick={[Function]}
    >
      <dl
        className="flex p-2 overflow-hidden relative"
      >
        <div
          className="flex flex-auto"
        >
          <dt
            className="
                duration-200
                flex-auto
                text-2xl font-bold text-warmGray-400
                pr-6
            "
          >
            Aceite de Oliva
          </dt>
        </div>
        <div
          className="
                absolute right-0
                pr-2
                bg-warmGray-100
                flex
                transform translate-x-24
                border border-solid border-white
                rounded-l-full
                duration-300
            "
        >
          <button
            className="
                    bg-white
                    icon-circle-left
                    flex items-center justify-center
                    text-warmGray-500
                    text-2xl
                    transform rotate-0
                    w-10
                    false
                    border border-solid border-white
                    duration-300
                "
            onClick={[Function]}
            type="button"
          />
          <div
            className=" grid grid-cols-2 gap-2 "
          >
            <button
              className=" icon-pencil px-2 text-2xl text-blue-500 "
              onClick={[Function]}
              type="button"
            />
            <button
              className=" icon-bin2 px-2 text-2xl text-rose-500 "
              onClick={[Function]}
              type="button"
            />
          </div>
        </div>
      </dl>
      <div
        className="
                        rounded-br-lg rounded-bl-lg
                        bg-warmGray-100
                        p-1
                    "
      >
        <ul
          className="flex flex-wrap"
        >
          <li
            className="
                            bg-white
                            text-xs text-warmGray-800
                            m-1 p-1
                        "
            key="500ml-0-function random() { [native code] }"
          >
            500ml
          </li>
          <li
            className="
                            bg-white
                            text-xs text-warmGray-800
                            m-1 p-1
                        "
            key="Carbonell-1-function random() { [native code] }"
          >
            Carbonell
          </li>
          <li
            className="
                            bg-white
                            text-xs text-warmGray-800
                            m-1 p-1
                        "
            key="metro-2-function random() { [native code] }"
          >
            metro
          </li>
          <li
            className="
                            bg-white
                            text-xs text-warmGray-800
                            m-1 p-1
                        "
            key="2 und-3-function random() { [native code] }"
          >
            2 und
          </li>
        </ul>
      </div>
    </div>
  </li>
  <li
    className="mb-3"
    key="Harina Pan-1-function random() { [native code] }"
  >
    <div
      className="
                duration-200
                rounded-lg undefined
                border border-solid border-warmGray-300
            "
      onClick={[Function]}
    >
      <dl
        className="flex p-2 overflow-hidden relative"
      >
        <div
          className="flex flex-auto"
        >
          <dt
            className="
                duration-200
                flex-auto
                text-2xl font-bold text-warmGray-400
                pr-6
            "
          >
            Harina Pan
          </dt>
        </div>
        <div
          className="
                absolute right-0
                pr-2
                bg-warmGray-100
                flex
                transform translate-x-24
                border border-solid border-white
                rounded-l-full
                duration-300
            "
        >
          <button
            className="
                    bg-white
                    icon-circle-left
                    flex items-center justify-center
                    text-warmGray-500
                    text-2xl
                    transform rotate-0
                    w-10
                    false
                    border border-solid border-white
                    duration-300
                "
            onClick={[Function]}
            type="button"
          />
          <div
            className=" grid grid-cols-2 gap-2 "
          >
            <button
              className=" icon-pencil px-2 text-2xl text-blue-500 "
              onClick={[Function]}
              type="button"
            />
            <button
              className=" icon-bin2 px-2 text-2xl text-rose-500 "
              onClick={[Function]}
              type="button"
            />
          </div>
        </div>
      </dl>
      <div
        className="
                        rounded-br-lg rounded-bl-lg
                        bg-warmGray-100
                        p-1
                    "
      >
        <ul
          className="flex flex-wrap"
        >
          <li
            className="
                            bg-white
                            text-xs text-warmGray-800
                            m-1 p-1
                        "
            key="1 Kg-0-function random() { [native code] }"
          >
            1 Kg
          </li>
          <li
            className="
                            bg-white
                            text-xs text-warmGray-800
                            m-1 p-1
                        "
            key="PAN-1-function random() { [native code] }"
          >
            PAN
          </li>
          <li
            className="
                            bg-white
                            text-xs text-warmGray-800
                            m-1 p-1
                        "
            key="San Felipe-2-function random() { [native code] }"
          >
            San Felipe
          </li>
          <li
            className="
                            bg-white
                            text-xs text-warmGray-800
                            m-1 p-1
                        "
            key="2 und-3-function random() { [native code] }"
          >
            2 und
          </li>
        </ul>
      </div>
    </div>
  </li>
</ul>
`;

exports[`Pruebas sobre archivo ListadoInsumos.js Se debe renderizar correctamente <ListadoInsumos /> 1`] = `
<h1
  className="font-bold text-2xl text-center text-warmGray-500"
>
  Debe agregrar insumos :)
</h1>
`;
